### ISSUE TITLE
Spotted a Predictable Anomaly with Incremental Order #'s

### ISSUE THREAD

**Issue by Dirty-Butter** [cubecart/v6#2439]
I've commented several times since Incremental's inception about rare odd number jumps in the sequence. I just realized what the pattern is, which I hope is meaningful and fixable. 

Here's a string of order #'s that occured in November: 
2019-PC-4543, 2019-PC-4544, **2019-PC-0753**, 2019-PC-4546

Now today I have  
2019-PC-4594, 2019-PC-4595, **2019-PC-0754**

Looking back through ALL orders since using the Incremental system - ALL orders with the 
**2019-PC -075x** consecutive series of numbers are for orders shipped to ALABAMA - MY Store home state, with MY Tax structure.

We ship very few orders to our own state, Alabama, but this is now a reproducible issue. I tried a test fake order, and got the same results.

**Comment by Dirty-Butter** [comment]
I noticed we had one order shipping to Alabama that occured right after a CC upgrade, and I did something that messed up Sales Tax and did not charge any sales tax on that order. THAT order follows the normal Increment sequence, 2019-PC-4484. So it may have to do with Sales Tax code????

**Comment by bhsmither** [comment]
A forum conversation is happening here:
https://forums.cubecart.com/topic/55681-duplicate-order-id-assigned-order-not-created-but-payment-taken/

**Comment by bhsmither** [comment]
Testing shows that the following `CREATE TRIGGER` statement gives much more (possibly 100%) reliable results.

My tests are against a MyISAM engine for the CubeCart_order_summary table. Hopefully, privileges issues* will not be an obstacle for an InnoDB engine table.

In /classes/order.class.php, near line 847:
```
$GLOBALS['db']->misc(
  "CREATE TRIGGER `custom_oid` BEFORE INSERT ON `".$GLOBALS['config']->get('config', 'dbprefix')."CubeCart_order_summary` FOR EACH ROW 
  SET NEW.custom_oid = "
  .str_replace(
    '`id`'
  , "(SELECT `auto_increment` FROM INFORMATION_SCHEMA.TABLES WHERE table_schema = '".$GLOBALS['config']->get('config', 'dbdatabase')."' AND table_name = '".$GLOBALS['config']->get('config', 'dbprefix')."CubeCart_order_summary')"
  , $concat
  )
);
```
(The PHP statement is broken apart for readability.)
* Privileges: https://dev.mysql.com/doc/refman/5.7/en/information-schema-introduction.html#information-schema-privileges

**Comment by Dirty-Butter** [comment]
Could you please show me the exact beginning/ ending lines to replace with your code @bhsmither ? I can test it with a fake Alabama order.

**Comment by bhsmither** [comment]
Note: I made an edit to the suggested code above. I accidentally used my database name instead of the config value.

**Comment by bhsmither** [comment]
Line 847 is the only line to change.
```
From:
$GLOBALS['db']->misc("CREATE TRIGGER `custom_oid` BEFORE INSERT ON `".$GLOBALS['config']->get('config', 'dbprefix')."CubeCart_order_summary` FOR EACH ROW SET NEW.custom_oid = ".str_replace('`id`', 'LAST_INSERT_ID()', $concat));

To:
$GLOBALS['db']->misc("CREATE TRIGGER `custom_oid` BEFORE INSERT ON `".$GLOBALS['config']->get('config', 'dbprefix')."CubeCart_order_summary` FOR EACH ROW SET NEW.custom_oid = ".str_replace('`id`', "(SELECT `auto_increment` FROM INFORMATION_SCHEMA.TABLES WHERE table_schema = '".$GLOBALS['config']->get('config', 'dbdatabase')."' AND table_name = '".$GLOBALS['config']->get('config', 'dbprefix')."CubeCart_order_summary')", $concat));
```
To cause the new trigger to take effect, the admin must simply Save the admin Store Settings. There is no need to actually change anything, just Save.

Note FYI: Every time the admin saves Store Settings, among other things, CubeCart will reconfigure the database to use a trigger to implement (or de-implement) the custom_oid feature. Every time!

In so doing the reconfiguration, the trigger is dropped and re-created, and all your email templates are scanned to update - replacing the 'other' cart ID identifier with the 'specified' cart ID identifier.

Which means you cannot be using `{$DATA.cart_order_id}` and `{$DATA.custom_oid}` template variables at the same time.

**Comment by Dirty-Butter** [comment]
I just tested your edit. It created a different number than the series of wrong order id's I've been getting (my sales tax ones have been sequential - just with their own series), but it did NOT pick up with the regular sequence it should have.

**Comment by Dirty-Butter** [comment]
I double checked to see what the Store Settings Preview of incremental said the next number should be - and **it matched the result of your code change**. So that looks like your edit worked. 

The problem with my install numbering must have something to do with the old plugin. If I use your code change and edit my Increment Increase Number in Store Settings, I just might fix it. 

But I won't test that theory until you say it's ok to try it.

**Comment by bhsmither** [comment]
In phpMyAdmin, execute the follow statement:
```
SELECT `auto_increment` FROM INFORMATION_SCHEMA.TABLES WHERE table_schema = 'myDatabaseName' AND table_name = 'prefix_CubeCart_order_summary';
```
(Replace `myDatabaseName` and `prefix_` as necessary.)

Note the number in the resultant table, `auto_increment` column.

Add to that value the number you have specified for the Increment Increase.

The total value is the number that should then form part of the entire Custom ID string that will be used on the next order -- regardless.

I say 'regardless' because the construction of the Custom ID happens exclusively in the database engine as the result of a trigger. However, there is always the possibility some extra code in CubeCart could be updating that column after the fact.

**Comment by Dirty-Butter** [comment]
I was never able to get the exact "next coid" number I wanted when I originally changed to CC's sequential. So it doesn't surprise me that this did not give me the next number in the sequence, either. I left if where it ended up and tried one test order with tax and one without. It still is not working for me. I give up.

**Comment by abrookbanks** [comment]
This is disappointing because Brians method seems like a really solid solution where was ask for a specific value and not just the last insert id from any table!

It seems to me like it could be taking the last insert ID from the tax table of you are charging tax to your home state.

**Comment by Dirty-Butter** [comment]
I would agree with that. You just have to remember that my pre-coid code was using an old SFWS Incremental Order Number plugin. I did everything I could think of to scrub any code or references to code out of my install and database - but maybe I was not successful and likely never will be now.

**Comment by Dirty-Butter** [comment]
I have also noticed the outlier from normal incremental numbering if the customer uses a coupon.

**Comment by abrookbanks** [comment]
There have been a number of issues and threads about this but I decided to open this one. 

We have to retire the trigger as it just doesn't work.

**Comment by Dirty-Butter** [comment]
My test code is on 6.4.1, and this new #2439 still makes one series if there is no tax or coupon and another series if there is sales tax or promotional code. Would there be anything in 6.4.2 commit that I should change as well?

**Comment by abrookbanks** [comment]
I forgot to say that it needs to be activated by clicking save in your stores settings. That will deactivate the trigger and enable to new code.

**Comment by Dirty-Butter** [comment]
First test worked! With and without sales tax were in sequence as they should be. And the predicted next number is actually what came next in the series - that had never happened. Will continue to test as soon as I can.

**Comment by Dirty-Butter** [comment]
Tested on 2 stores at 6.4.1 and it worked on both of them with and without coupons or sales tax. YAY!!

**Comment by abrookbanks** [comment]
Great!!

**Comment by Dirty-Butter** [comment]
One last piece of information about custom_oid. When I upgraded plushmemories/lovie-shoppe I could not get the Preview of the custom id number to show what I expected. I had forgotten what it took to get the dirtybutterestates.com custom_oid to work. The "old" increment increase number would not erase from the database config file even when I blanked it out and saved in CC Admin. I had to decode the database config table, delete the old increment increase number and save the encoded config again. THEN I could set the increment increase to the number needed to get the next order number to be correct.
SO the last used increment increase number is not wiped out of the database when someone experiments with incremental.
